### PR TITLE
Modified GetHashCode and Equals methods to support default(StringSegm…

### PR DESCRIPTION
**Pull request by jbagga**
…ent) comparison. Added tests.

### CODE REVIEW

**Comment by jbagga** [comment]
cc @pranavkm

File: test/Microsoft.Extensions.Primitives.Tests/StringSegmentTest.cs [hunk 1]
```diff
@@ -230,10 +230,13 @@ public static TheoryData GetHashCode_ReturnsSameValueForEqualSubstringsData
             {
                 return new TheoryData<StringSegment, StringSegment>
                 {
+                    { default(StringSegment), default(StringSegment) },
+                    { default(StringSegment), new StringSegment() },
                     { new StringSegment("Test123", 0, 0), new StringSegment(string.Empty) },
                     { new StringSegment("C`est si bon", 2, 3), new StringSegment("Yesterday", 1, 3) },
                     { new StringSegment("Hello", 1, 4), new StringSegment("Hello world", 1, 4) },
-                    { new StringSegment("Hello"), new StringSegment("Hello", 0, 5) },
+                    { new StringSegment("Hello"), new StringSegment("Hello", 0, 5) }
```
Comment (pranavkm) [inline test/Microsoft.Extensions.Primitives.Tests/StringSegmentTest.cs]: Formatting: Need a trailing comma here and remove the extra line that follows.

File: test/Microsoft.Extensions.Primitives.Tests/StringSegmentTest.cs [hunk 2]
```diff
@@ -260,7 +263,7 @@ public static TheoryData GetHashCode_ReturnsDifferentValuesForInequalSubstringsD
                     { new StringSegment(testString, 0, 1), new StringSegment(string.Empty) },
                     { new StringSegment(testString, 0, 1), new StringSegment(testString, 1, 1) },
                     { new StringSegment(testString, 1, 2), new StringSegment(testString, 1, 3) },
-                    { new StringSegment(testString, 0, 4), new StringSegment("TEST123", 0, 4) },
+                    { new StringSegment(testString, 0, 4), new StringSegment("TEST123", 0, 4) }
```
Comment (pranavkm) [inline test/Microsoft.Extensions.Primitives.Tests/StringSegmentTest.cs]: Undo this change.

File: test/Microsoft.Extensions.Primitives.Tests/StringSegmentTest.cs [hunk 3]
```diff
@@ -292,6 +295,61 @@ public void StringSegment_EqualsString_Invalid()
             Assert.False(result);
         }
 
+        public static TheoryData<StringSegment, StringComparison, bool> DefaultEqualsStringSegmentData
```
Comment (pranavkm) [inline test/Microsoft.Extensions.Primitives.Tests/StringSegmentTest.cs]: `DefaultStringEqualsStringSegmentData`?

File: test/Microsoft.Extensions.Primitives.Tests/StringSegmentTest.cs [hunk 4]
```diff
@@ -292,6 +295,61 @@ public void StringSegment_EqualsString_Invalid()
...
+        {
+            get
+            {
+                // candidate / comparer / expected result
+                return new TheoryData<StringSegment, StringComparison, bool>()
+                {
+                    { default(StringSegment), StringComparison.Ordinal, true },
```
Comment (pranavkm) [inline test/Microsoft.Extensions.Primitives.Tests/StringSegmentTest.cs]: Other tests might do this, but it's much nicer to split out the `equals` and `not equals` scenarios in to two separate tests. In this case, move the last scenario `{ new StringSegment("Hello, World!", 1, 4), StringComparison.Ordinal, false }` to a separate test along the lines of `StringSegmentDoesNotEqualNonDefaultStringSegments`. This should change the test data to `TheoryData<StringSegment>`

File: test/Microsoft.Extensions.Primitives.Tests/StringSegmentTest.cs [hunk 5]
```diff
@@ -292,6 +295,61 @@ public void StringSegment_EqualsString_Invalid()
...
+            var segment = default(StringSegment);
+           
+            // Act
+            var result = segment.Equals(candidate, StringComparison.Ordinal);
+
+            // Assert
+            Assert.Equal(expectedResult, result);
+        }
+
+        public static TheoryData<string, StringComparison, bool> DefaultEqualsStringData
```
Comment (pranavkm) [inline test/Microsoft.Extensions.Primitives.Tests/StringSegmentTest.cs]: Same as above. These should be `DefaultStringSegment_DoesNotEqualStringsData`

File: test/Microsoft.Extensions.Primitives.Tests/StringSegmentTest.cs (hunk 3 above)
Comment (jbagga) [inline test/Microsoft.Extensions.Primitives.Tests/StringSegmentTest.cs]: DefaultStringSegmentEqualsStringSegmentData? I want to make a distinction between type string and StringSegment.

**Comment by jbagga** [comment]
cc @pranavkm Just made the changes requested

File: src/Microsoft.Extensions.Primitives/StringSegment.cs [hunk 6]
```diff
@@ -129,7 +129,7 @@ public bool Equals(StringSegment other)
         public bool Equals(StringSegment other, StringComparison comparisonType)
         {
             int textLength = other.Length;
-            if (!HasValue || Length != textLength)
+            if (Length != textLength)
```
Comment (pranavkm) [inline src/Microsoft.Extensions.Primitives/StringSegment.cs]: 😢 Trailing whitespace

File: test/Microsoft.Extensions.Primitives.Tests/StringSegmentTest.cs [hunk 7]
```diff
@@ -292,6 +294,74 @@ public void StringSegment_EqualsString_Invalid()
             Assert.False(result);
         }
 
+        public static TheoryData<StringSegment, StringComparison> DefaultStringSegmentEqualsStringSegmentData
+        {
+            get
+            {
+                // candidate / comparer / expected result
```
Comment (pranavkm) [inline test/Microsoft.Extensions.Primitives.Tests/StringSegmentTest.cs]: Unless you want to add variants of `StringComparison.OrdinalIgnoreCase` you don't need to pass it to the test. Also, 
// candidate / comparer ~~/ expected result~~

File: test/Microsoft.Extensions.Primitives.Tests/StringSegmentTest.cs [hunk 8]
```diff
@@ -292,6 +294,74 @@ public void StringSegment_EqualsString_Invalid()
...
+                {
+                    { string.Empty, StringComparison.Ordinal, false },
+                    { "Hello, World!", StringComparison.Ordinal, false },
+                };
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(DefaultStringSegmentDoesNotEqualStringData))]       
+        public void DefaultStringSegment_DoesNotEqualString(string candidate, StringComparison comparison, bool expectedResult)
```
Comment (pranavkm) [inline test/Microsoft.Extensions.Primitives.Tests/StringSegmentTest.cs]: You don't need `expectedResult` so remove it from the `TheoryData`

File: test/Microsoft.Extensions.Primitives.Tests/StringSegmentTest.cs [hunk 9]
```diff
@@ -292,6 +294,74 @@ public void StringSegment_EqualsString_Invalid()
...
+           
+            // Act
+            var result = segment.Equals(candidate, StringComparison.Ordinal);
+
+            // Assert
+            Assert.True(result);
+        }
+
+        [Fact]
+        public void DefaultStringSegment_DoesNotEqualStringSegment()
```
Comment (pranavkm) [inline test/Microsoft.Extensions.Primitives.Tests/StringSegmentTest.cs]: Could you convert this into a theory and add a test data with `new StringSegment(string.Empty)` and `new StringSegment("Hello", 1, 0)`?

File: test/Microsoft.Extensions.Primitives.Tests/StringSegmentTest.cs (hunk 8 above)
Comment (pranavkm) [inline test/Microsoft.Extensions.Primitives.Tests/StringSegmentTest.cs]: Also, you aren't using `comparison`.

**Comment by jbagga** [comment]
cc @pranavkm Incorporated changes as requested
